### ISSUE TITLE
project search by keywords

### ISSUE THREAD

**Issue by clairerye** [ebi-ait/dcp-ingest-central#386]
As wrangler I want to find specific projects so that I can organise my work

Change the search to allow users to select partial or full word search with and/or logic on multiple search terms, and search across all fields in the metadata schema

** Acceptance Criteria / Definition of Done **

- [x] Given an identifier for a project - e.g. UUID, Accession (GEO/AE etc), DOI, publication link/title the matching project will be identified
- [x] Given a search of a metadata field eg. technology, organ, disease, protocol, all matching projects will be identified
- [x] Given a search of "liver" all projects where liver is either mentioned in the title etc or is listed as the organ should show in the results, user can indicate if "full word search" is required in order to remove answers such as "Oliver"
- [x] Given a search of "liver" and "Teichmann" all projects that contain both terms somewhere in the metadata should show in the results
- [ ] Given the publication title or the project title the matching project will be identified

*stretch goal - (probably a separate task)*
- [ ] search also related entities, e.g. biomaterials, diseases

**Comment by amnonkhen** [comment]
[Full text search in mongo](https://docs.mongodb.com/manual/text-search/) can be used here.
It can be done on all textual fields or just on some.

It supports:
- searching for a list of strings in any order, e.g. `liver Teichmann`
- searching for a phrase, e.g. `"left lung"`

not sure about `liver` vs. `oliver`

**Comment by jacobwindsor** [comment]
I have moved the UUID searching into #393 since it is not so simple but not too difficult. Just to get the other features into prod more quickly

**Comment by clairerye** [comment]
waiting on #313

**Comment by clairerye** [comment]
Exact match search is needed. 
Here are some examples that we would like to work more effectively.
searching for "Single-cell transcriptome profiling of the vaginal wall in women with severe anterior vaginal prolapse" in ingest UI gives 533 results (from a total of 541 possible)
Only 1 project in ingest exactly matches that and we would expect to see that one project at the top of the search list.

**Comment by amnonkhen** [comment]
The scenario Claire described on Aug 3 looks like "any keyword" query, which is the default of mongo text index.
I believe this should be changed to an "all keyword" query.
The questions are:
- [ ] Do we only allow "all keywords" queries?
- [ ] Do we let the user choose between "any keyword", "all keywords", "exact match"?

@gabsie what do you think?

**Comment by Wkt8** [comment]
Possibly being demo'd today. @jacobwindsor and @amnonkhen to discuss progress

**Comment by amnonkhen** [comment]
There is a new form field for defining the search type. The user can choose between all keywords, any keyword or an exact phrase match. 

See screen shot:


![image.png](https://images.zenhubusercontent.com/600ed297091d3ea152b83c2d/68b9bf85-84a9-4188-99ad-fefb30fd8abe)

**Comment by Wkt8** [comment]
Doing some testing.

**Comment by Wkt8** [comment]
Search Type: 
- All Keywords
- Any Keywords
- Exact Match

What's the difference between all keywords and any keywords? 

Have tested the following: 
UUID (does not work) 
ProjectShortName (works with some bugs)
- Given Project ID: HeterogeneityCD4Tcells
- Searching on Search Type 'Any Keyword' 
- Search: Heterogeneity - works
- Search: CD4 - works
- Search: HeterogeneityCD4 - does not work??? 

Disease (not working(low priority)) 
PMID (no, but not included in scope) 

Accession (works) 
Name (works)
Project Description (works) 
Author (works) 
DOI (works)
Publication link (no) 
Technology (works)
Organ (works) 
Species (works) 
Contributor (works)

I think allowing searches by uuid would be useful, and was also included in the original scope of the ticket. Also there are a couple bugs in searching ProjectShortName.

**Comment by gabsie** [comment]
Hi, @amnonkhen and @Wkt8 

From a user perspective, I would expect some sort of a broad search and get rid of the box for exact match etc, etc.. These options should be incorporated into the system/code, but maybe not visualised in an additional box.
 
The system I expect to display at the top the exact matches.. then the rest if relevant

To me the highest priority is still to search by short name, and by UUID - if exact matches they should just display those results?

The way this currently works, it should have the button for Search I think.
I will incorporate this in the design together with the review of the filters, as discussed on the [other ticket.](https://app.zenhub.com/workspaces/dcp-ingest-product-development-5f71ca62a3cb47326bdc1b5c/issues/ebi-ait/dcp-ingest-central/450)

Gabs

**Comment by amnonkhen** [comment]
search by uuid - ticket 398
search by shot name - TBD

**Comment by ofanobilbao** [comment]
Tested already
